### PR TITLE
Clearup warnings

### PR DESCRIPTION
**Pull request by eddiebergman**
This PR attempts to clear up some warnings that appear in the Github actions [tests](https://github.com/automl/auto-sklearn/runs/3537886422?check_suite_focus=true). Warnings that could be considered harmful are left in and documented here.

This PR description will updated as progress is made and sources of the warnings identified.

Some of the warnings are easy fixes and just warnings about future API changes. Other warnings we are aware of and ignore such as for IterativeAlgorithms in which they don't converge as we are manually iterating.

# Potential problem warnings:
* Illuminates underlying issue that we don't consider datasets with unique labels throughout the codebase 
```
 test/test_evaluation/test_train_evaluator.py::TestTrainEvaluator::test_datasets
test/test_evaluation/test_train_evaluator.py::TestTrainEvaluator::test_datasets
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/sklearn/model_selection/_split.py:666: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=2.
    warnings.warn(("The least populated class in y has only %d"
```

* Uncertain what this means for metafeatures
```

test/test_metalearning/pyMetaLearn/test_meta_features.py::test_calculate_all_metafeatures_multilabel[pandas]
test/test_metalearning/pyMetaLearn/test_meta_features.py::test_calculate_all_metafeatures_multilabel[numpy]
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/numpy/lib/nanfunctions.py:1670: RuntimeWarning: Degrees of freedom <= 0 for slice.
    var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,

test/test_metalearning/pyMetaLearn/test_meta_features.py::test_calculate_all_metafeatures_multilabel[pandas]
test/test_metalearning/pyMetaLearn/test_meta_features.py::test_calculate_all_metafeatures_multilabel[numpy]
  /home/runner/work/auto-sklearn/auto-sklearn/autosklearn/metalearning/metafeatures/metafeatures.py:474: RuntimeWarning: Mean of empty slice
    mean = np.nanmean(values)
```

* QDA might be being used incorrectly
```
test/test_pipeline/test_classification.py: 3 warnings
test/test_pipeline/components/classification/test_qda.py: 36 warnings
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/sklearn/discriminant_analysis.py:808: UserWarning: Variables are collinear
    warnings.warn("Variables are collinear")
```

* Clustering metrics may be being misused, the continuous labels are likely just floats such as `1.0` or `0.0` as labels but the warning seems to indicate something doesn't like multiclass problems. 
```
test/test_pipeline/test_classification.py: 51 warnings
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/sklearn/metrics/cluster/_supervised.py:58: UserWarning: Clustering metrics expects discrete values but received continuous values for label, and multiclass values for target
    warnings.warn(msg, UserWarning)
```

* It seems two or three regression algorithms are producing large numbers somewhere and causing overflows which probably ruins their performance. Converting these tests to be pytests with fixtures would help illuminate which algorithms are causing these overflows.
```
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations_signed_data
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations_signed_data
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations_signed_data
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/numpy/core/_methods.py:233: RuntimeWarning: overflow encountered in multiply
    x = um.multiply(x, x, out=x)

test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations_signed_data
test/test_pipeline/test_regression.py::SimpleRegressionPipelineTest::test_configurations_signed_data
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/numpy/core/_methods.py:244: RuntimeWarning: overflow encountered in reduce
    ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)
```

* LDA seems to be getting something it doesn't like during it's fitting process
```
test/test_pipeline/components/classification/test_lda.py::LDAComponentTest::test_module_idempotent
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/sklearn/discriminant_analysis.py:480: RuntimeWarning: invalid value encountered in true_divide
    self.explained_variance_ratio_ = (S**2 / np.sum(
```

* Same with some feature processor
```
test/test_pipeline/components/feature_preprocessing/test_select_rates_regression.py::SelectRegressionRatesComponentTest::test_default_configuration
test/test_pipeline/components/feature_preprocessing/test_select_rates_regression.py::SelectRegressionRatesComponentTest::test_default_configuration
  /usr/share/miniconda/envs/testenv/lib/python3.8/site-packages/sklearn/feature_selection/_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
    corr /= X_norms
```

* More invalid values encountered
```
test/test_pipeline/test_classification.py::SimpleClassificationPipelineTest::test_configurations_signed_data
  /opt/hostedtoolcache/Python/3.7.11/x64/lib/python3.7/site-packages/sklearn/decomposition/_fastica.py:60: RuntimeWarning: invalid value encountered in sqrt
    return np.linalg.multi_dot([u * (1. / np.sqrt(s)), u.T, W])
```

### CODE REVIEW

File: autosklearn/pipeline/components/base.py [hunk 1]
```diff
@@ -145,12 +147,21 @@ def __str__(self):
 
 class IterativeComponent(AutoSklearnComponent):
     def fit(self, X, y, sample_weight=None):
-        self.iterative_fit(X, y, n_iter=2, refit=True)
-        iteration = 2
-        while not self.configuration_fully_fitted():
-            n_iter = int(2 ** iteration / 2)
-            self.iterative_fit(X, y, n_iter=n_iter, refit=False)
-            iteration += 1
+        with warnings.catch_warnings():
```
Comment (mfeurer) [inline autosklearn/pipeline/components/base.py:150]: Is this to not have this warning shown in the unit tests?

File: autosklearn/pipeline/components/base.py [hunk 2]
```diff
@@ -162,15 +173,27 @@ def get_current_iter(self):
...
     def fit(self, X, y, sample_weight=None):
-        self.iterative_fit(
-            X, y, n_iter=2, refit=True, sample_weight=sample_weight
-        )
-        iteration = 2
-        while not self.configuration_fully_fitted():
-            n_iter = int(2 ** iteration / 2)
-            self.iterative_fit(X, y, n_iter=n_iter, sample_weight=sample_weight)
-            iteration += 1
+        with warnings.catch_warnings():
```
Comment (mfeurer) [inline autosklearn/pipeline/components/base.py:178]: Same here.

File: autosklearn/metalearning/metalearning/meta_base.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def add_dataset(self, name, metafeatures):
             self.metafeatures.drop(name.lower(), inplace=True)
         self.metafeatures = self.metafeatures.append(metafeatures)
 
-        runs = pd.Series([], name=name)
+        runs = pd.Series([], name=name, dtype=object)
```
Comment (mfeurer) [inline autosklearn/metalearning/metalearning/meta_base.py:68]: Looking at this, I'm currently wondering whether this should a) be a dataframe as the other entries in `self.algorithm_runs`, and b) be of type float

**Review by mfeurer — COMMENTED** [review]
Ha, that was too quick again, sorry, I need to check whether you already requested a review next time.

File: autosklearn/pipeline/components/base.py (hunk 1 above)
Comment (eddiebergman) [reply autosklearn/pipeline/components/base.py:150]: Well more in general but yes it also is for the unittests, I'm not sure where these warnings get swallowed in general, perhaps to the log file. We are aware that iterative algorithms will usually complain about convergence due to how we do things so I thought it would be best to consume them.

Thought's on alternatives:
1. Leave this specific warning to be omitted?
2. Other?

**Comment by eddiebergman** [comment]
> Ha, that was too quick again, sorry, I need to check whether you already requested a review next time.

It's fine, this PR will likely need a few rounds of discussion on how best to handle some warnings.

File: autosklearn/pipeline/components/base.py (hunk 1 above)
Comment (mfeurer) [reply autosklearn/pipeline/components/base.py:150]: Yes, they are redirected to the logfile in this function: https://github.com/automl/auto-sklearn/blob/62a66731b13be2b1c4002c232d89e4d3b211d268/autosklearn/evaluation/abstract_evaluator.py#L153

For that reason I'd prefer to let them reach the logfile instead. Would it be possible to swallow these in the unit tests? I think we can also remove quite a few warnings by updating `autosklearn.pipeline.components.base:IterativeComponent.fit` to be
```
    def fit(self, X, y, sample_weight=None):
        self.iterative_fit(X, y, n_iter=self.get_max_iter(), refit=True)
        return self
```
which as a nice side-effect could make Auto-sklearn as a whole faster (because stuff like GradientBoosting doesn't have to do initialization on calls to fit).

Comment (eddiebergman) [reply autosklearn/pipeline/components/base.py]: We actually get different results if we do it this way, the assertion errors for explicit values change (minorly).

File: autosklearn/metalearning/metalearning/meta_base.py (hunk 3 above)
Comment (eddiebergman) [reply autosklearn/metalearning/metalearning/meta_base.py:68]: It's a Series of type float.

File: autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py [hunk 4]
```diff
@@ -28,21 +29,25 @@ def fit(self, X: PIPELINE_DATA_DTYPE,
...
             # We do not want to impute a category with the default
-            # value (0 is the default) in case such default is in the
-            # train data already!
-            fill_value = 0
-            unique = np.unique(X)
-            while fill_value in unique:
-                fill_value -= 1
+            # value (0 is the default).
+            # Hence we take one greater than the max
+            unique = np.unique([*X.data, 0]) if issparse(X) else np.unique(X)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: `*X.data` looks like it could be a really costly operation. Could we do `np.unique(X.data)` here and then `fill_value = min(min(unique) - 1, 0)`?

File: autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py [hunk 5]
```diff
@@ -28,21 +29,25 @@ def fit(self, X: PIPELINE_DATA_DTYPE,
...
+            print(unique)
+            fill_value = min(unique) - 1
+        else:
+            fill_value = None
+
+        print(fill_value)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:50]: Please drop the two print statements.

File: test/test_pipeline/components/data_preprocessing/test_categorical_imputation.py [hunk 6]
```diff
@@ -34,15 +34,16 @@ def test_default_imputation(input_data_imputation, categorical):
...
         X[mask] = np.nan
     else:
-        imputation_value = 0
+        imputation_value = min(np.unique(X)) - 1
+
     Y = CategoricalImputation().fit_transform(X.copy())
-    assert ((np.argwhere(Y == imputation_value) == np.argwhere(mask)).all())
-    assert ((np.argwhere(Y != imputation_value) == np.argwhere(np.logical_not(mask))).all())
+
+    assert np.array_equal(Y == imputation_value, mask)
```
Comment (mfeurer) [inline test/test_pipeline/components/data_preprocessing/test_categorical_imputation.py:41]: How about np.testing.assert_array_equal ?

**Review by mfeurer — COMMENTED** [review]
Looks mostly good to me (two minor comments), but the tests are failing right now :(

File: autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py (hunk 4 above)
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: That's a better idea, I wasn't really thinking about any memory movement.

File: autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py (hunk 5 above)
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:50]: Whoops

File: test/test_pipeline/components/data_preprocessing/test_categorical_imputation.py (hunk 6 above)
Comment (eddiebergman) [reply test/test_pipeline/components/data_preprocessing/test_categorical_imputation.py:41]: Sure why not, probably gives better feedback on failure

File: autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py (hunk 4 above)
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: That tests fail due to having a negative categorical label, I'll switch to max instead.

Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: ... and switching to max seems to break other tests

Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: It seems like fixing this is going to take quite a while, otherwise we wait for using the `OrdinalEncoder` to take care of this. For now, I'll default it back to the original silent error that sparse matrices have `nans` encoded to 0.

Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: Turns out, reverting back to the orignal code now gives tests failures -_-

Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: This [loc](https://github.com/automl/auto-sklearn/blob/master/autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py#L43) fails for sparse matrices yet I have no idea how it still works in the current master branch?

Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py:44]: Okay never mind it doesn't but it doesn't work as intended
```python
x = scipy.sparse.csr_matrix([[0,0,1],[0,1,0]])
unique - np.unique(x)
print(unique)

array([<2x3 sparse matrix of type '<class 'numpy.int64'>'
       	with 3 stored elements in Compressed Sparse Row format>],
      dtype=object)
      
In [10]: 0 in np.unique(x)
/home/skantify/.pyenv/versions/3.8.5/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3444: SparseEfficiencyWarning: Comparing a sparse matrix with 0 using == is inefficient, try using != instead.
  exec(code_obj, self.user_global_ns, self.user_ns)
<ipython-input-10-73ac6367503d>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  0 in np.unique(x)
  
Out[10]: False

In [11]: 1 in np.unique(x)

<ipython-input-11-053263455b76>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  1 in np.unique(x)
  
Out[11]: False
```

**Review by mfeurer — APPROVED** [review]

File: autosklearn/pipeline/components/data_preprocessing/categorical_encoding/encoding.py [hunk 7]
```diff
@@ -27,7 +27,14 @@ def fit(self, X: PIPELINE_DATA_DTYPE,
                 categories='auto', handle_unknown='use_encoded_value', unknown_value=-1,
             )
             self.preprocessor.fit(X, y)
-        return self
+            return self
+        else:
+            # TODO spare_encoding of negative labels
```
Comment (mfeurer) [inline autosklearn/pipeline/components/data_preprocessing/categorical_encoding/encoding.py:32]:
```suggestion
            # TODO sparse_encoding of negative labels
```

**Review by mfeurer — APPROVED** [review]
